warc_metadata_sidecar.py

```diff
@@ -0,0 +1,261 @@
+    hostname = socket.gethostname()
+    warc_doc = 'http://bibnum.bnf.fr/WARC/WARC_ISO_28500_version1_latestdraft.pdf'
+    warcinfo_payload = {'software': 'warc-metadata-sidecar/' + version,
+                        'hostname': hostname,
+                        'ip': socket.gethostbyname(hostname),
+                        'conformsTo': warc_doc,
+                        'description': 'WARC metdata sidecar for ' + new_file,
+                        'publisher': publisher}
+    if operator:
+        warcinfo_payload['operator'] = operator
+    if 'isPartOf' in warc_list:
+        list_index = warc_list.index('isPartOf')
+        is_part_of = warc_list[list_index + 1]
+        warcinfo_payload['isPartOf'] = is_part_of
+
+    return warcinfo_payload
+
+
+def metadata_sidecar(archive_dir, warc_file, operator=None,
+                     publisher='University of North Texas - Digital Projects Unit'):
+    start = time.time()
+    version = pkg_resources.require("warc-metadata-sidecar")[0].version
+
+    if not os.path.isdir(archive_dir):
+        os.mkdir(archive_dir)
+
+    logging.basicConfig(
+        filename=os.path.join(archive_dir, 'sidecar.log'),
+        level=logging.INFO,
+        format='%(asctime)s %(levelname)s %(message)s',
+    )
+    logging.getLogger(__name__)
+    logging.info('Logging WARC metadata record information for %s', warc_file)
+
+    # Create sidecar filename, adding 'meta' as extension.
+    new_file = os.path.basename(warc_file)
+    meta_file = re.sub('warc(\.gz)?$', 'warc.meta.gz', new_file)
+    logging.info('Creating sidecar %s', meta_file)
+    warc_file_path = os.path.join(archive_dir, meta_file)
+
+    fido = ExtendFido()
+    mime_title = 'Identified-Payload-Type:'
+    puid_title = 'Preservation-Identifier:'
+    charset_title = 'Charset-Detected:'
+    language_title = 'Languages-cld2:'
+
+    # Open the sidecar file to write in the metadata, open the warc file to get each record.
+    with open(warc_file_path, 'ab') as output, open(warc_file, 'rb') as stream:
+        record_count = 0
+        total_records = 0
+        text_mime = 0
+        non_text = 0
+        non_mime = 0
+        for record in ArchiveIterator(stream):
+            total_records += 1
+            url = record.rec_headers.get_header('WARC-Target-URI')
+            payload = io.BytesIO(record.content_stream().read())
+            if not record_count:  # and record.rec_type == 'warcinfo':
+                rec_type = record.rec_type
+                warc_info = create_warcinfo_payload(payload, operator, publisher,
+                                                    new_file, version)
+                # Write warcinfo record into sidecar.
+                writer = WARCWriter(output, gzip=False)  # TODO: gzip will equal True
+                warcinfo_record = writer.create_warcinfo_record(meta_file, warc_info)
+                writer.write_record(warcinfo_record)
+                record_count += 1
```

Comment:
Maybe this warcinfo creating code should be moved to above the `for record in ArchiveIterator(stream)` section, because we only want to do it once per WARC input file. Also, I am not sure if it should be counted in `record_count` if that is for tracking records that are found and processed.

Reply:
I have been avoiding this long enough, but I finally made an attempt to moving this above the for loop. I wasn't sure how to get just the one record, because warcio doesn't have code for that. Then I remembered that it's just an open file. 😀 But I do think you may have a nicer solution than what I came up with.

```
if '.gz' in warc_file:
     stream = gzip.GzipFile(fileobj=stream)
lines = stream.readlines(1000)
for line in lines:
     is_partof = re.search(r'\bisPartOf: +([^\r\n]+)', line.decode('utf-8'))
     if is_partof:
          break
warc_info = create_warcinfo_payload(is_partof, new_file, version,
                                                    operator, publisher)
# Write warcinfo record into sidecar.
warcinfo_record = writer.create_warcinfo_record(meta_file, warc_info)
writer.write_record(warcinfo_record)
```
If you know of a different/better way, I'm all ears.

Reply:
err when I commented I had forgotten we were using part of the original WARC's possibly present isPartOf field in the new warcinfo record and was focusing on how we only want to write one warcinfo record to the new file. We'll need to think about whether we want to bother with the isPartOf field or not in the sidecar and/or what we want to do if there is a warcinfo that is not at the top of the file :/.